### ISSUE TITLE
FR: contextual-interface

### ISSUE THREAD

**Issue by rafaelss95** [angular-eslint/angular-eslint#455]
Currently we have a rule to check if a lifecycle is in the correct context:

[**`contextual-lifecycle`**](https://github.com/angular-eslint/angular-eslint/blob/master/packages/eslint-plugin/src/rules/contextual-lifecycle.ts)

But in reality, there are many other interfaces that I would like to see being checked. Here is the complete list that I managed to put together:

| Full list |
|:-:|
| CanActivate |
| CanActivateChild |
| CanDeactivate |
| CanLoad |
| Resolve |
| HttpInterceptor |
| PipeTransform |
| ControlValueAccessor |
| OnChanges |
| OnInit |
| DoCheck |
| AfterContentInit |
| AfterContentChecked |
| AfterViewInit |
| AfterViewChecked |
| OnDestroy |
| DoBootstrap |

That being said, I'd like to propose a rule to enforce all these interfaces (and probably replace the `contextual-lifecycle`).
Note that this rule could also accept an optional configuration to let people creation its custom configuration. An example could be like this:

```js
{
  "rules": {
    "@angular-eslint/contextual-interface": [
      "error",
      {
        "OnInitEffects": {
          "decorators": [
            "Injectable"
          ]
        },
        "MyInterface": {
          "decorators": [
            "Component", "Directive"
          ]
        }
      }
    ]
  }
}
```

Note that this config shouldn't replace the default, but merge.
